prow/gerrit/adapter/trigger.go

```diff
@@ -36,15 +40,35 @@ func messageFilter(lastUpdate time.Time, change client.ChangeInfo, presubmits []
 			continue
 		}
 
-		if !pjutil.TestAllRe.MatchString(message.Message) {
+		if pjutil.TestAllRe.MatchString(message.Message) {
+			filters = append(filters, pjutil.TestAllFilter())
+		} else if pjutil.RetestRe.MatchString(message.Message) {
+			if latestReport == nil {
+				continue
+			}
+			allContexts := sets.String{}
+			failedContexts := sets.String{}
+
+			jobs := map[string]*reporter.Job{}
+			for _, job := range latestReport.Jobs {
+				jobs[job.Name] = job
+			}
+			for _, presubmit := range presubmits {
+				allContexts.Insert(presubmit.Name)
+				j, ok := jobs[presubmit.Name]
+				if ok && strings.ToLower(j.State) == string(v1.FailureState) {
+					failedContexts.Insert(presubmit.Name)
+				}
+			}
+			filters = append(filters, pjutil.RetestFilter(failedContexts, allContexts))
```

Comment:
This is starting to look a lot like the logic in `generic-comment.go#presubmitFilter()` -- can that be DRY with this?

Reply:
They have different ways of getting the status contexts though, and also the ok-to-test thing above.

Reply:
So factor those two out and pass them in as parameters?

Reply:
I think if we want to have the same UX here need a DRY method that opens up the ways that the two differ

Reply:
Done.